pkg/controllers/proxy/service_endpoints_sync.go

```diff
@@ -0,0 +1,517 @@
+		glog.Errorf("Error setting up IPVS services for service external IP's and load balancer IP's: %s", err.Error())
+	}
+	err = nsc.cleanupStaleVIPs(activeServiceEndpointMap)
+	if err != nil {
+		syncErrors = true
+		glog.Errorf("Error cleaning up stale VIP's configured on the dummy interface: %s", err.Error())
+	}
+	err = nsc.cleanupStaleIPVSConfig(activeServiceEndpointMap)
+	if err != nil {
+		syncErrors = true
+		glog.Errorf("Error cleaning up stale IPVS services and servers: %s", err.Error())
+	}
+	err = nsc.syncIpvsFirewall()
+	if err != nil {
+		syncErrors = true
+		glog.Errorf("Error syncing ipvs svc iptables rules to permit traffic to service VIP's: %s", err.Error())
+	}
+	err = nsc.setupForDSR(serviceInfoMap)
+	if err != nil {
+		syncErrors = true
+		glog.Errorf("Error setting up necessary policy based routing configuration needed for direct server return: %s", err.Error())
+	}
+
+	if syncErrors {
+		glog.V(1).Info("One or more errors encountered during sync of IPVS services and servers to desired state")
+	} else {
+		glog.V(1).Info("IPVS servers and services are synced to desired state")
+	}
+
+	return nil
+}
+
+func (nsc *NetworkServicesController) setupClusterIPServices(serviceInfoMap serviceInfoMap, endpointsInfoMap endpointsInfoMap, activeServiceEndpointMap map[string][]string) error {
+	ipvsSvcs, err := nsc.ln.ipvsGetServices()
+	if err != nil {
+		return errors.New("Failed get list of IPVS services due to: " + err.Error())
+	}
+	for k, svc := range serviceInfoMap {
+		var protocol uint16
+
+		switch svc.protocol {
+		case "tcp":
+			protocol = syscall.IPPROTO_TCP
+		case "udp":
+			protocol = syscall.IPPROTO_UDP
+		default:
+			protocol = syscall.IPPROTO_NONE
+		}
+
+		endpoints := endpointsInfoMap[k]
+		dummyVipInterface, err := nsc.ln.getKubeDummyInterface()
+		if err != nil {
+			return errors.New("Failed creating dummy interface: " + err.Error())
+		}
+		// assign cluster IP of the service to the dummy interface so that its routable from the pod's on the node
+		err = nsc.ln.ipAddrAdd(dummyVipInterface, svc.clusterIP.String(), true)
+		if err != nil {
+			continue
+		}
+
+		// create IPVS service for the service to be exposed through the cluster ip
+		ipvsClusterVipSvc, err := nsc.ln.ipvsAddService(ipvsSvcs, svc.clusterIP, protocol, uint16(svc.port), svc.sessionAffinity, svc.scheduler, svc.flags)
+		if err != nil {
+			glog.Errorf("Failed to create ipvs service for cluster ip: %s", err.Error())
+			continue
+		}
+		var clusterServiceId = generateIpPortId(svc.clusterIP.String(), svc.protocol, strconv.Itoa(svc.port))
+		activeServiceEndpointMap[clusterServiceId] = make([]string, 0)
+
+		// add IPVS remote server to the IPVS service
+		for _, endpoint := range endpoints {
+			dst := ipvs.Destination{
+				Address:       net.ParseIP(endpoint.ip),
+				AddressFamily: syscall.AF_INET,
+				Port:          uint16(endpoint.port),
+				Weight:        1,
+			}
+
+			err := nsc.ln.ipvsAddServer(ipvsClusterVipSvc, &dst)
```

Comment:
@murali-reddy I tested this and overall it looks good as it concerns ClusterIP services. The only problem I found is that if the service is a local service and it has active endpoints on the node, then this code still adds all of the endpoints to the IPVS service which means that service local services could still end up with an extra hop instead of staying on the host.

If you add the following code block right above this line then it mitigates this and keeps service local traffic on the node:
```
			// Conditions on which to add an endpoint on this node:
			// 1) Service is not a local service
			// 2) Service is a local service, but has no active endpoints on this node
			// 3) Service is a local service, has active endpoints on this node, and this endpoint is one of them
			if svc.local {
				if hasActiveEndpoints(svc, endpoints) && !endpoint.isLocal {
					continue
				}
			}
```

Reply:
@aauren thanks for the review. I have updated the PR